### ISSUE TITLE
[PR #164] [WIP] doc: added proposal specs for model_provider, oagw and cred_store components

### ISSUE THREAD

**Issue by orivaris** [constructorfabric/gears-rust#2156]
> 🔗 **Mirrored PR** [cyberfabric/cyberware-rust#164](https://github.com/cyberfabric/cyberware-rust/pull/164) | **Author:** Artifizer | **Opened:** 2025-12-27T01:22:15Z | **Status:** closed (not merged)
> *GitHub API does not allow setting PR author or timestamps — attribution preserved here.*

---

This is some sort of requirements for future services development/generation as part of preparation for Chat development. Here we prepare foundation for secure communication with external Model Providers, such as OpenAI, Anthropic, etc. The Models, Chat, File Parser and local Search will come next 

---

**Comment by orivaris** [comment]
**codecov-commenter** commented on 2025-12-27T19:46:43Z:

---

:warning: Please install the !['codecov app svg image'](https://github.com/codecov/engineering-team/assets/152432831/e90313f4-9d3a-4b63-8b54-cfe14e7ec20d) to ensure uploads and comments are reliably processed by Codecov.

## [Codecov](https://app.codecov.io/gh/hypernetix/hyperspot/pull/164?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=None) Report
:white_check_mark: All modified and coverable lines are covered by tests.

:loudspeaker: Thoughts on this report? [Let us know!](https://github.com/codecov/feedback/issues/255)

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/oagw/spec.md` line 294 on 2025-12-27T08:22:28Z:

---

Labels tenant_id, route_id, link_id are high cardinality metrics. Normally, it's not recommended to use them, as they blow up production monitoring infrastructure and get blacklisted.

**Comment by orivaris** [comment]
**Artifizer** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/oagw/spec.md` line 294 on 2025-12-27T09:09:59Z:

---

agree, however Victoria Metrics deal with it much better than Prometheus, so we can simulate it and benchmark it before making final decision

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/oagw/spec.md` line 296 on 2025-12-27T10:06:27Z:

---

For prometheus metrics it is idiomatic to add suffix with measurement units like `duration_seconds` Also for counter suffix `_total` is idiomatic

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/oagw/spec.md` line 303 on 2025-12-27T10:13:07Z:

---

Wouldn't it be better and more reliable to have counters with hits_total, misses_total and rate can be just derived from counter?

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/oagw/spec.md` line 322 on 2025-12-27T10:16:55Z:

---

Does it mean if k8s tries to probe /health endpoint it needs to get auth token? For me it doesn't sound good.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 31 on 2025-12-27T10:41:00Z:

---

In real systems DB credentials provided by external systems and consuming from k8s secrets directly.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 55 on 2025-12-27T10:43:25Z:

---

I think the purpose of secret versioning is not clear.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 193 on 2025-12-27T10:47:23Z:

---

probably for sqlite we should use this https://www.sqlite.org/see/doc/trunk/www/readme.wiki

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 4 on 2025-12-27T11:02:05Z:

---

`lowest priority` is ambiguous and contradicts later statements where “lower wins”. Let's use one unambiguous rule: priority is an integer, lower value wins. I sugget to replace this phrase with 'single active plugin chosen by minimal priority value'.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 170 on 2025-12-27T11:06:15Z:

---

This implies plugins store the ciphertext as the source of truth. Right now you are designing two owners for the same data. Decide the ownership model:

Either gateway owns encryption and stores ciphertext, plugin is thin storage.

OR plugin owns encryption and stores ciphertext, gateway stores only metadata and plugin references.
Do not do both unless you want double-encryption and split-brain.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 96 on 2025-12-27T11:08:26Z:

---

Looks like  "explicitly authorized retrieval modes" is undefined

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 93 on 2025-12-27T11:10:28Z:

---

If GET returns plaintext, it will a leak  (as I understand we don't return plain secrets) 
If GET returns metadata, you need a separate reveal endpoint. Right now this is ambiguous.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 241 on 2025-12-27T11:15:48Z:

---

This interacts with the gateway vs plugin ownership decision. If Hashicorp Vault is used, the DB should store only a reference, not ciphertext. If Transit is used, define where ciphertext lives and who calls Transit. Right now both interpretations are possible.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 263 on 2025-12-27T11:18:17Z:

---

If you pass ciphertext into the plugin and then write to KV, you are storing ciphertext as the "secret value" and will need client-side decrypt, which defeats the purpose of Vault KV. 

If using Vault KV, pass plaintext to plugin and let Vault handle encryption-at-rest

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 319 on 2025-12-27T11:20:07Z:

---

Let's define a single policy model with explicit precedence: global default, per-tenant override, per-secret-type override and etc. Right now this reads like two unrelated rules.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 614 on 2025-12-27T11:21:12Z:

---

Do not log kek_id if it can help attackers identify key material or rotation state

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 621 on 2025-12-27T11:22:41Z:

---

Need to address that write operations MUST be idempotent in case of retries

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/model-provider/spec.md` line 550 on 2025-12-27T12:31:24Z:

---

Should this really be a service-level requirement scenario? It looks more like a requirement for observability/monitoring platform.

**Comment by orivaris** [comment]
**ast2074** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 364 on 2025-12-27T15:05:10Z:

---

should be separate roles for accessing the metadata and accessing the secret itself

**Comment by orivaris** [comment]
**ast2074** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 103 on 2025-12-27T15:08:22Z:

---

i'd suggest to consider only plugins for encryption, not for storage: 1) encryption by hs vault in transit mode, 2) custom own encryption impl

**Comment by orivaris** [comment]
**ast2074** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 277 on 2025-12-27T15:10:02Z:

---

i guess, per-module audit will be implemented same way for any module - so that's a good room for generalization and moving into the framework

**Comment by orivaris** [comment]
**ast2074** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 440 on 2025-12-27T15:12:57Z:

---

i'd suggest to generalize all compliance-related (gdpr, hippa, pci/dss etc.) requirements, they all are about few ideas actually, the main ones are audit (incl. retention) and rotating keks

**Comment by orivaris** [comment]
**ast2074** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 1 on 2025-12-27T15:18:42Z:

---

general comments: 
- no current/target overall architecture view (or at least main architectural ideas) available, it's hard to 'deduce' it from a set of 'requirements'
- too many requirements at once, they surely need to be prioritized and divided by phases/milestones

**Comment by orivaris** [comment]
**Artifizer** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/oagw/spec.md` line 303 on 2025-12-27T23:45:29Z:

---

OK

**Comment by orivaris** [comment]
**Artifizer** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/oagw/spec.md` line 322 on 2025-12-27T23:51:40Z:

---

OK for health and probe, but I would keep a role for /oagw/v1/routes/{routeId}/health still

**Comment by orivaris** [comment]
**Artifizer** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/oagw/spec.md` line 296 on 2025-12-27T23:53:54Z:

---

OK

**Comment by orivaris** [comment]
**Artifizer** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/model-provider/spec.md` line 550 on 2025-12-27T23:54:51Z:

---

Agree

**Comment by orivaris** [comment]
**Artifizer** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/cred-store/spec.md` line 4 on 2025-12-27T23:55:49Z:

---

OK

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/model-provider/spec.md` line 419 on 2025-12-28T10:00:56Z:

---

I think it's api_ingress/api gateway responsibility

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/model-provider/spec.md` line 1 on 2025-12-28T10:03:14Z:

---

Basically not clear who is an owner of HTTP call to a model:

1. model_provider
2. oagw
3. plugin

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/model-provider/spec.md` line 555 on 2025-12-28T10:05:17Z:

---

Looks like this call doesn't support all model selection criteries:

- capabilities
- features
- tenant requirements
- quotas
- priority

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/model-provider/spec.md` line 236 on 2025-12-28T10:07:55Z:

---

It will be a too hot table in case of high QPS. Probably we need to implement async audi logging via some message queue or use some time series (prometheus?)

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `openspec/changes/add-model-provider-cred-store-oagw/specs/model-provider/spec.md` line 1 on 2025-12-28T10:13:27Z:

---

Moreover: what exactly is the ownership and granularity of plugins?

The spec currently does not clearly define what a plugin represents:
- a plugin per concrete LLM provider (e.g. OpenAI, Anthropic),
- a plugin per API/interface style (chat/completions, embeddings, streaming, etc.),
- or a plugin per concrete model family / endpoint variant.

This ambiguity affects:
- plugin selection logic (what `can_handle()` is really matching),
- capability/feature matching semantics,
- caching strategy (client cache key correctness),
- future extensibility (e.g. multiple APIs per vendor, mixed transports, local vs remote runtimes).

Suggestion:
- Explicitly define plugin ownership boundaries, for example:
  - plugins are responsible for a `(vendor + API style + transport)` combination, and models are data/config within providers, or
  - plugins are responsible for a `(vendor + model family)` and expose supported capabilities/endpoints declaratively.
- Document this as an invariant, not just an example, so both gateway logic and plugin authors have a single mental model.

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 76 on 2025-12-28T19:44:36Z:

---

IMO OpenSpec is very bad tool if we stick to "use your favourite tool approach. In such case it's better to simulate OpenSpec ad-hoc without its injected prompts and AGENTS.md polluting context.

OpenSpec relies on ./openspec/spec to by always in-sync with current code. It's impossible unless every commit that adds new features or significantly changes behavior is done with OpenSpec flow.

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 134 on 2025-12-28T19:52:49Z:

---

Is there FACTS_TAXONOMY file in repository?

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/docs/1_INTENT_DESIGN.md` line 3 on 2025-12-29T12:08:32Z:

---

This looks like it's not properly formatted:
```suggestion
See [oagw/spec.md](@/openspec/changes/add-model-provider-cred-store-oagw/specs/oagw/spec.md)
```

**Comment by orivaris** [comment]
**Bechma** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 19 on 2025-12-29T12:10:52Z:

---

I feel like we need to specify the "@" symbol to notate that we are talking about the content root, and not the relative path.

**Comment by orivaris** [comment]
**Bechma** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 76 on 2025-12-29T12:13:19Z:

---

I agree, we need to specify the tool to use or a list of tools.

**Comment by orivaris** [comment]
**Bechma** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 105 on 2025-12-29T12:14:22Z:

---

criticize based on what?

**Comment by orivaris** [comment]
**Bechma** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 54 on 2025-12-29T12:20:29Z:

---

what is v2+? It needs clarification

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/domain/service.rs` line 61 on 2025-12-29T12:23:58Z:

---

Construction of policy_engine inside new violates DI principles. Dependency should be passed via constructor.

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/Cargo.toml` line 27 on 2025-12-29T12:24:16Z:

---

no need to specify features that are already on the top-level

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/api/rest/dto.rs` line 209 on 2025-12-29T12:27:00Z:

---

Why? we already have axum::http or http package

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/api/rest/error.rs` line 9 on 2025-12-29T12:29:51Z:

---

why this implementation lives here? Very confusing, should live in the domain::error

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/domain/service.rs` line 268 on 2025-12-29T12:32:41Z:

---

This is cutting corner for v1. 

suggestion: it would be helpful to have ACTUAL_vs_INTENT on the stage of implementation PR like this, so that these kind of simplifications become transparent

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/api/rest/handlers.rs` line 60 on 2025-12-29T12:33:11Z:

---

do we want to support $select?

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/api/rest/routes.rs` line 96 on 2025-12-29T12:38:07Z:

---

Why are we exposing two handlers that do the same? By convention there's the health endpoint to check the liveness of the services

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/api/rest/routes.rs` line 279 on 2025-12-29T12:41:46Z:

---

this is not how we do it, please, refer to the example:
https://github.com/hypernetix/hyperspot/blob/main/examples/modkit/users_info/users_info/src/api/rest/routes.rs#L91

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/api/rest/routes.rs` line 280 on 2025-12-29T12:42:08Z:

---

same situation, please, refer to example:
https://github.com/hypernetix/hyperspot/blob/main/examples/modkit/users_info/users_info/src/api/rest/routes.rs#L93

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/domain/ports.rs` line 12 on 2025-12-29T12:45:23Z:

---

Why we have this trait here? If this is a placeholder for credstore... fine, but looks like it should reside in the modkit-security

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/domain/ports.rs` line 43 on 2025-12-29T12:46:50Z:

---

this type is a stub also? if not, we should be use a const

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/domain/service.rs` line 47 on 2025-12-29T12:51:19Z:

---

This is a violation of DI and DDD principles. 

ClientHub is a struct, not a trait, and it does not belong to domain. And basically such injection says that Service depends on every client that might possible exist. Injecting ClientHub is basically infamours ServiceLocator antipattern.

A better approach would be for Service to depend on something like oagw-specific plugin registry trait. 
An implementation of this trait (infrastructure) is an adapter around ClientHub.

With current approach how would one unit-test business logic in domain service?

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/domain/repo.rs` line 1 on 2025-12-29T12:51:32Z:

---

Can someone explain why we need traits here? For testing purposes?

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/domain/service.rs` line 51 on 2025-12-29T12:54:04Z:

---

In current form Service does too many things:
- CRUD of routes
- CRUD of links
- link resolution
- Invokations

These are separate concerns

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/infra/storage/entity.rs` line 1 on 2025-12-29T12:54:08Z:

---

What is the reason of suddenly encapsulate models inside of inner modules?

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/infra/storage/route_repo.rs` line 43 on 2025-12-29T13:01:34Z:

---

According to clippy.toml we should 
`Use SecureSelect::all() after .secure().scope_with(scope)`

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/infra/storage/route_repo.rs` line 165 on 2025-12-29T13:02:23Z:

---

Access should be controlled by AccessScope

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/infra/storage/route_repo.rs` line 196 on 2025-12-29T13:02:56Z:

---

Access should be controlled by AccessScope

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/infra/storage/route_repo.rs` line 208 on 2025-12-29T13:03:14Z:

---

Access should be controlled by AccessScope

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/lib.rs` line 4 on 2025-12-29T13:04:11Z:

---

that's not correct, consistency means to use `to_owned` everywhere as this is how we approached in all the project.
We must attach to the project structure or drop the structure in favour of the definition of "consistency" for each person.

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-sdk/src/gts.rs` line 14 on 2025-12-29T13:12:54Z:

---

Why are we using mod here? what are we trying to achieve?

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-sdk/src/models.rs` line 16 on 2025-12-29T13:15:48Z:

---

This enum was already created somewhere I reviewed.

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-sdk/src/models.rs` line 30 on 2025-12-29T13:16:49Z:

---

This is not correct, you need to implement AsStr trait:
https://docs.rs/enum_str/latest/enum_str/trait.AsStr.html
pub trait AsStr {
    // Required method
    fn as_str(&self) -> &str;
}

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-sdk/src/models.rs` line 90 on 2025-12-29T13:17:31Z:

---

Why I feel like a dejavu? I've seen this file somewhere... -.-

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/api/rest/dto.rs` line 92 on 2025-12-29T13:55:26Z:

---

This definitely requires unit-test that Some(None) and None are serialized/deserialized differently.

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/api/rest/dto.rs` line 113 on 2025-12-29T13:57:11Z:

---

Didn't we discuss using `snake_case` for DTOs?

**Comment by orivaris** [comment]
**hyphen-2025** reviewed `modules/oagw/oagw-gw/src/domain/repo.rs` line 1 on 2025-12-29T14:07:58Z:

---

It's DDD/hexagonal architecture port and adapters concept. I guess somewhere in the prompts, we specified that we use DDD. Domain service depends on ports (interfaces/traits) defined in domain module (not in outside modules) and does not depend on external layers. External layers depend on domain and are responsible for integrations via adapters.

It's helpful for injecting mocks, and sometimes helpful for complicated domains to have more disciplined approach to dependencies and coupling.

**Comment by orivaris** [comment]
**aviator5** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 1 on 2025-12-29T15:41:24Z:

---

`CODE_GEN_PIPELINE.md` is currently in `guidelines/` dir but it's actually a human workflow for module developing and orchestrating LLMs, not guidelines that AI agents should follow. Suggest having a separate `workflows` dir and place this file there. This keeps `guidelines/` focused on norms/patterns that both humans and AI agents should follow (NEW_MODULE.md, SECURITY.md, etc.).

I believe we will implement tools that will automate pipelines from the `workflows/` dir and that will instruct AI agents to use guidelines from the `guidelines/` dir.

Additionally, in the `workflow/` we may have `prompts` sub-dir with workflow-related prompts and reference to them in the workflow markdown files.

Also, I would suggest renaming `CODE_GEN_PIPELINE.md` to `GEN_NEW_MODULE_PIPELINE.md`. It's more concrete, not about about abstract code.

**Comment by orivaris** [comment]
**aviator5** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 6 on 2025-12-29T15:42:11Z:

---

"LLM-actionable" confuses a bit - it's a pipeline for humans, not instructions for AI agents.

**Comment by orivaris** [comment]
**Bechma** reviewed `modules/oagw/oagw-gw/src/api/rest/dto.rs` line 92 on 2025-12-29T16:50:50Z:

---

I can't think about a valid Option<Option case.
the lint error is correct, the allow statement is incorrect

**Comment by orivaris** [comment]
**aviator5** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 136 on 2025-12-29T17:49:53Z:

---

Do we really need to version scenarios or implementation phases? In fact, to avoid ambiguities I would suggest align terms in the following way:
1. Each **workflow** (please see my [previous comment](https://github.com/hypernetix/hyperspot/pull/164/files#r2651236615) for the details) consists of multiple **steps** (not **phases**).
2. `INTENT_DESIGN.md` describes architecture of the modules, its dependencies, implementations details, etc. All those things are divided into **phases** (phase - is a more familiar word here in my view). Naming: `<MODULE_PREFIX>-P<NUMBER>` (for better grepping and cross-referencing): `OAGW-P1`, `CREDSTORE-P42`, etc. I would suggest introducing the strict rule: every design aspect (authorization, tenancy isolation, licensing support, etc.) must have list of features divided into phases and some phase-to-phase migrations notes.
4. `INTENT_SCENARIOS.md` contains scenarios  aligned with **phases** from the `INTENT_DESIGN.md`. So, for example, `modules/oagw/docs/3_INTENT_SCENARIOS.md` may have two "[OAGW-P1] Timeout Configuration" and "[OAGW-P2] Timeout Configuration" which may be similar or completely different. Propose to **introduce ID for each scenario** (`<MODULE_PREFIX>-S<NUMBER>`): `OAGW-S1`, `CREDSTORE-S100500`, etc.
5. `INTENT_FACTS.md` contains facts (I believe we need to brainstorm to invent better name for this entity) aligned with **phases** from the `INTENT_DESIGN.md`. Approach is the same as for scenarios. Identifiers (`<MODULE_PREFIX>-F<NUMBER>`): `OAGW-F73`, etc.
6. Each module has **version** (version of the rust crate). I would suggest having `CHANGELOG.md` file that will
    - follow https://keepachangelog.com/en/1.0.0/ format
    - contain info what scenarios (human-readable phrase + ID) was added/fixed in which version.
Example:
```
  ## [0.2.0] - 2024-01-15
  ### Added
  - OAGW-S5: Timeout configuration support
  - OAGW-S12: Retry with exponential backoff
 ```

**UPD**: Suggest using **REQUIREMENT** word instead of **FACT**: authorization requirements, tracing requirements, logging requirements, etc. It looks better name for reusable implementation details referenced from scenarios ("Use OAGW-R42 for tenancy-isolation check").

**UPD2**: Alternatives in naming: `PHASE-OAGW-42`, `REQ-OAGW-7`, `SCEN-OAGW-10`.

**Comment by orivaris** [comment]
**aviator5** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 105 on 2025-12-29T18:05:54Z:

---

I think it's just an concept, of course we need more comprehensive prompts on each step.

**Comment by orivaris** [comment]
**aviator5** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 76 on 2025-12-29T18:12:27Z:

---

I'm not sure that we need to use any existing SDD tool for now at least for generating `DESIGN.md`, `FACTS.md`, `SCENARIOS.md`, etc.

**Comment by orivaris** [comment]
**aviator5** reviewed `modules/oagw/oagw-gw/src/config.rs` line 13 on 2025-12-29T18:19:40Z:

---

How about using `std::time::Duration` with `#[serde(with = "humantime_serde")]` attribute? I will allow having human readable "10s", "1h" in config.

**Comment by orivaris** [comment]
**aviator5** reviewed `modules/oagw/oagw-gw/src/config.rs` line 37 on 2025-12-29T18:23:48Z:

---

Do we plan to support rate limits with granularity finer than “per minute” (e.g., 10 per sec)?
If yes, could we make the configuration more flexible-something like `10/s`, `100/m`, etc. - instead of hardcoding minute-based limits?

**Comment by orivaris** [comment]
**aviator5** reviewed `modules/oagw/oagw-gw/src/config.rs` line 51 on 2025-12-29T18:24:31Z:

---

`max_audit_latency_ms` -> `audit_max_latency_ms`?
for better consistency

**Comment by orivaris** [comment]
**aviator5** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 19 on 2025-12-29T18:40:16Z:

---

Probably worth converting our guidelines into [Agent Skills](https://github.com/agentskills/agentskills) (the open standard maintained by Anthropic). It’s gaining traction as the next “big” standard alongside/after MCP.

It’s already supported (or landing) in multiple tools:
- Claude Code
- OpenAI Codex CLI: https://developers.openai.com/codex/skills/
- GitHub Copilot: https://docs.github.com/en/copilot/concepts/agents/about-agent-skills
- Cursor (nightly): https://cursor.com/docs/context/skills

Looks like skills support is a future for AI Agents for proper context management.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/module.rs` line 63 on 2025-12-29T19:57:09Z:

---

Accrosing to intent scenarios (5.1 Gateway Module Initialization), we should register:
gts.x.core.oagw.proto.v1~
gts.x.core.oagw.stream_proto.v1~
gts.x.core.oagw.auth_type.v1~
and Load well-known instances from gts/*.instances.json.

But I don't see it here.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/Cargo.toml` line 1 on 2025-12-29T19:58:25Z:

---

Boundaries and module parts locations looks ok.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/module.rs` line 91 on 2025-12-29T20:01:43Z:

---

I think it's not ok. We should pass ModuleCtx everywhere.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/config.rs` line 13 on 2025-12-29T20:06:04Z:

---

I think it should be std::time::Duration with humantime representation in yaml config.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/domain/ports.rs` line 12 on 2025-12-29T20:10:54Z:

---

Do we really need t oabstract secret resolver via ports pattern? I thought that we can call Credstore (whatever) client here.

Anyway ports in domain layer should be traits not concrete implementations. (https://github.com/hypernetix/hyperspot/blob/main/examples/modkit/users_info/users_info/src/domain/ports/audit.rs) 

The specific port implementation should be implemented in src/infra (https://github.com/hypernetix/hyperspot/blob/main/examples/modkit/users_info/users_info/src/infra/audit/http_audit_client.rs)

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/api/rest/routes.rs` line 38 on 2025-12-29T20:12:52Z:

---

I think it's redudant. Anyway it's not used in code:
```Rust
.operation_id("oagw.health")
```

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-sdk/src/api.rs` line 134 on 2025-12-29T20:15:16Z:

---

I suggest to use XXXClient suffix. It's really client for some API but not API itself.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/module.rs` line 31 on 2025-12-29T20:23:28Z:

---

Dependency on a specific plugin module (oagw_default_plugin) weakens “pluggable” architecture.
It is fine as a dev default, but production should not require a single named plugin.

Let's remove this dependency.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/api/rest/routes.rs` line 346 on 2025-12-29T20:33:01Z:

---

I think we will have too many hops here. Let's imagine completion call:
Vendor client GET /v1/chat/completions -> API Gateway -> LLM Inference module (?) -> API Gateway -> OAGW -> OAGW Plugin -> real API (OpenAI API for ex.)

In case of Out Of Process modules the number of hops is increasing.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/infra/storage/link_repo.rs` line 63 on 2025-12-29T20:38:27Z:

---

If we want to have pagination here this method is wrong:
1. Wrong signature. Please check: https://github.com/Artifizer/hyperspot/blob/06cb74af0d049ca1275effdd774700db9fae15b8/examples/modkit/users_info/users_info/src/infra/storage/sea_orm_repo.rs#L190

2. Wrong returning result: https://github.com/Artifizer/hyperspot/blob/06cb74af0d049ca1275effdd774700db9fae15b8/examples/modkit/users_info/users_info/src/infra/storage/sea_orm_repo.rs#L203C7-L214C15

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/infra/storage/link_repo.rs` line 1 on 2025-12-29T20:39:40Z:

---

I don't understand why do we need this repo. Links are a part of common data model of OAGW. Let's implemnt all these links operations in main repo.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/infra/storage/entity.rs` line 1 on 2025-12-29T20:48:18Z:

---

This layout (multiple SeaORM entities defined as submodules in a single file) is technically valid and will work correctly with SeaORM. BUT!

### File organization

* Defining multiple entities in a single file is acceptable, but diverges from the typical SeaORM convention (one entity per file)
* This may complicate future use of `sea-orm-cli generate entity` and increase merge conflicts as the schema grows
* Consider splitting entities into separate files if the module expands like:
   1. The file entities/mod.rs for re-export
   2. route, link, route_protocol: separate files

### Cross-tenant consistency between `link` and `route`

* `link` stores both `tenant_id` and `route_id`, but there is no DB-level constraint ensuring `link.tenant_id == route.tenant_id`.
* This invariant must be enforced at the service layer; otherwise, it is possible to create a link referencing a route owned by a different tenant.
* Recommendation: add explicit validation on create/update paths and cover this with tests.

### Join table primary key design

* `route_protocol` uses a surrogate `id` as the primary key.
* This allows duplicate `(route_id, protocol_gts_id)` rows unless a unique constraint exists.
* Recommendation: add a unique constraint on `(route_id, protocol_gts_id)` to preserve logical integrity.

### Use of raw `String` for GTS identifiers

* Fields such as `auth_type_gts_id`, `strategy_gts_id`, and `protocol_gts_id` are modeled as `String`.
* This weakens type safety and makes misconfiguration possible at runtime.
* Recommendation: We already have a typed wrapper like `GtsInstanceId`

**Comment by orivaris** [comment]
**Artifizer** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 105 on 2025-12-29T21:38:07Z:

---

Yes, this is a conceptual placeholder that will be refined later; however modern models already handle such loosely defined tasks quite well.

**Comment by orivaris** [comment]
**Artifizer** reviewed `guidelines/CODE_GEN_PIPELINE.md` line 134 on 2025-12-29T21:38:24Z:

---

No, it needs to be added

**Comment by orivaris** [comment]
**Artifizer** reviewed `modules/oagw/oagw-gw/src/api/rest/dto.rs` line 209 on 2025-12-29T21:41:32Z:

---

I believe it's because Dto must be a part of contract and dependency agnostic

**Comment by orivaris** [comment]
**Artifizer** reviewed `modules/oagw/oagw-gw/src/api/rest/error.rs` line 9 on 2025-12-29T21:48:17Z:

---

&#2369;MikeFalcon77 what do you think?

**Comment by orivaris** [comment]
**Artifizer** reviewed `modules/oagw/oagw-gw/src/api/rest/handlers.rs` line 60 on 2025-12-29T21:48:56Z:

---

Yes, in the future versions

**Comment by orivaris** [comment]
**Artifizer** reviewed `modules/oagw/oagw-gw/src/api/rest/routes.rs` line 96 on 2025-12-29T21:56:27Z:

---

In Kubernetes, readiness and liveness (“health”) probes serve different, very intentional purposes. Mixing them up causes real production issues, so this distinction matters.

Readiness: Can the pod handle traffic right now?
If it fails, Kubernetes stops sending traffic but does not restart the pod. This code will be implemented in v2 as it's mentioned in the "handlers::ready" handler comments

Liveness (health): Is the pod still alive?
If it fails, Kubernetes restarts the pod.

So we can use readiness for temporary or external issues; use liveness only when the process is broken and needs a restart.

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/oagw/oagw-gw/src/domain/service.rs` line 47 on 2025-12-31T12:25:10Z:

---

Yes, good catch. We should not inject ClientHub into domain/services.

The service should be assembled from explicit dependencies (ports), e.g. API client traits, repositories, publishers.

ClientHub is fine as an infrastructure-level registry, but it must be used only during wiring/composition (module init/runtime bootstrap) to resolve those ports and pass them into the service.

If we want stricter DDD, we can formalize ports in the domain layer and implement thin adapters in the infra layer that delegate to the actual API clients (local or gRPC). But I think it's too much for most our cases.
